### PR TITLE
Add contracts to Deroll

### PR DESCRIPTION
**Pull request by sandhilt**
Some features for Deroll
- ERC721
- ERC1155Single
- ERC1155Batch
- Add tests

### CODE REVIEW

**Comment by tuler** [comment]
Please organize the commits in a way that helps the review process.

**Comment by sandhilt** [comment]
> Please organize the commits in a way that helps the review process.

Ok. I will rebase that. Thanks for quickly response.

**Comment by sandhilt** [comment]
Hello @tuler, please feel free for send comments and review this PR. If need more fixes, please, let me know.

File: packages/wallet/build-local.sh [hunk 1]
```diff
@@ -0,0 +1,7 @@
+npm run prepack
+npm pack
+cp deroll-wallet-0.3.7.tgz ../../../cartesify-nodejs-rest-example/backend
+
+cd ../../../cartesify-nodejs-rest-example/backend
```
Comment (tuler) [inline packages/wallet/build-local.sh:5]: this should not be here.
or the entire `build-local.sh` file

File: packages/wallet/smart_contracts/erc1155-abi.json
Comment (tuler) [inline packages/wallet/smart_contracts/erc1155-abi.json:1]: rename to `packages/wallet/abi/erc1155.ts` and return the abi `as const`, so it integrates well with `abitype` and `viem`.

File: packages/wallet/src/index.ts [hunk 2]
```diff
@@ -9,8 +9,14 @@ import {
...
-import { erc20PortalAddress, etherPortalAddress } from "./rollups";
-
+import {
+    erc20PortalAddress,
+    etherPortalAddress,
+    erc721PortalAddress,
+    erc1155SinglePortalAddress,
+    erc1155BatchPortalAddress,
+} from "./rollups";
+import { decodeAbiParameters, parseAbiParameters } from "viem";
```
Comment (tuler) [inline packages/wallet/src/index.ts:19]: put these together with the other `viem` import in the beginning

File: packages/wallet/src/index.ts [hunk 3]
```diff
@@ -61,8 +87,83 @@ export const parseERC20Deposit = (payload: Payload): ERC20Deposit => {
     return { success, token, sender, amount };
 };
 
+/**
+ * Decode input according to https://github.com/cartesi/rollups/tree/v1.0.0#input-encodings-for-deposits
```
Comment (tuler) [inline packages/wallet/src/index.ts:91]: link to `1.2.0` just for consistency with the others

File: packages/wallet/src/index.ts [hunk 4]
```diff
@@ -61,8 +87,83 @@ export const parseERC20Deposit = (payload: Payload): ERC20Deposit => {
...
+
+    return {
+        token,
+        sender,
+        tokenId,
+    };
+};
+
+/**
+ * Decode input according to https://github.com/cartesi/rollups/tree/v1.0.0#input-encodings-for-deposits
```
Comment (tuler) [inline packages/wallet/src/index.ts:112]: link to `1.2.0` just for consistency with the others

File: packages/wallet/src/token.ts [hunk 5]
```diff
@@ -0,0 +1,95 @@
+import { isValidAdvanceRequestData } from "./util";
+import { InvalidPayloadError } from "./errors";
+import { getAddress, type Address, type Hex } from "viem";
```
Comment (tuler) [inline packages/wallet/src/token.ts:3]: put in the beginning

**Comment by tuler** [comment]
package `@types/node` is missing in devDependencies of wallet package

File: packages/wallet/src/wallet.ts [hunk 6]
```diff
@@ -1,28 +1,25 @@
...
-} from ".";
+import { Address, getAddress, isAddress } from "viem";
+
+import { inspect } from "node:util";
+import { TokenHandler } from "./token";
 
 export type Wallet = {
     ether: bigint;
-    erc20: Record<Address, bigint>;
+    erc20: Map<Address, bigint>;
```
Comment (tuler) [inline packages/wallet/src/wallet.ts:9]: Record is more appropriate than Map. Any reason of changing to Map?

File: packages/wallet/src/wallet.ts [hunk 7]
```diff
@@ -1,28 +1,25 @@
...
+    erc721: Map<Address, Set<bigint>>;
+    erc1155: Map<Address, Map<bigint, bigint>>;
 };
 
 export interface WalletApp {
     balanceOf(address: string): bigint;
     balanceOf(token: Address, address: string): bigint;
```
Comment (tuler) [inline packages/wallet/src/wallet.ts:16]: Maybe rename to `balanceOfERC20` for consistency

File: packages/wallet/src/wallet.ts [hunk 8]
```diff
@@ -31,91 +28,176 @@ export interface WalletApp {
...
+        });
+    }
 
-        if (walletFrom.ether < amount) {
-            throw new Error(`insufficient balance of user ${from}`);
+    public handler: AdvanceRequestHandler = async (data) => {
+        try {
+            console.log("Wallet handler...", inspect(data, { depth: null }));
+
+            const tokenHandler = TokenHandler.getInstance();
```
Comment (tuler) [inline packages/wallet/src/wallet.ts:170]: this `TokenHandler.getInstance` is weird. Why?

File: packages/wallet/src/wallet.ts [hunk 9]
```diff
@@ -31,91 +28,176 @@ export interface WalletApp {
...
+            owner,
+            getWallet: this.getWalletOrNew,
+        });
+    }
 
-        if (walletFrom.ether < amount) {
-            throw new Error(`insufficient balance of user ${from}`);
+    public handler: AdvanceRequestHandler = async (data) => {
+        try {
+            console.log("Wallet handler...", inspect(data, { depth: null }));
```
Comment (tuler) [inline packages/wallet/src/wallet.ts:168]: remove all console.log

**Comment by tuler** [comment]
I see a bunch of typescript errors

File: packages/wallet/src/wallet.ts (hunk 6 above)
Comment (sandhilt) [reply packages/wallet/src/wallet.ts:9]: Changing to maintain consistency in other tokens, Like ERC721 and ERC1155 that use Map and Set.

File: packages/wallet/src/wallet.ts (hunk 7 above)
Comment (sandhilt) [reply packages/wallet/src/wallet.ts:16]: `balanceof` is still using the current method that verifies the Ether or ERC20 balance.I can change too.

File: packages/wallet/src/wallet.ts (hunk 8 above)
Comment (sandhilt) [reply packages/wallet/src/wallet.ts:170]: The token or contract `Tokenhandler` is a singleton with a facade for each of the items: ERC20, Ether, ERC1155 (Single/Batch), ERC721.
It is useful in using the `handler` method where he find which token or contract should be used and executed.
In other methods, it is called the instance of this `Tokenhandler`.

Comment (tuler) [reply packages/wallet/src/wallet.ts:170]: I’d rather not use this pattern and just use the handler pattern that deroll itself uses.

**Comment by sandhilt** [comment]
Hello @tuler, I change some things based on your comments and adjust with @fabiooshiro. Please, let me know if it's better or needs more adjustments.

File: packages/wallet/src/errors.ts [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+export class InvalidPayloadError extends Error {
+    constructor(public payload: unknown) {
+        super(`Invalid payload: ${inspect(payload)}`);
+    }
+}
+
+export class RelayError extends Error {
+    constructor() {
+        super(
+            `You need to call the method relayDAppAddress from DAppAddressRelay__factory.`,
```
Comment (tuler) [inline packages/wallet/src/errors.ts:12]:
```suggestion
            `application has not received its address from DAppAddressRelay`,
```

File: packages/wallet/src/errors.ts [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+            `insufficient balance of user ${user} of token ${token}: ${amount.toString()}${
+                tokenId ? ` of tokenId ${tokenId}` : ""
+            }`,
+        );
+    }
+}
+
+export class WalletUndefinedError extends Error {
+    constructor(address: string) {
+        super(`wallet of user ${address} is undefined`);
```
Comment (tuler) [inline packages/wallet/src/errors.ts:29]: are we raising this in what situation? Isn't a wallet always defined in the perspective of the user, even with a zero balance.

File: packages/wallet/src/errors.ts [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+    }
+}
+
+export class TokenFromUserNotFound extends Error {
+    constructor(user: string, token: string, tokenId: bigint) {
+        super(`user ${user} does not have token ${tokenId} of token ${token}`);
```
Comment (tuler) [inline packages/wallet/src/errors.ts:35]: are we raising this in what situation? if a user doesn't have it, he just doesn't have it.

File: packages/wallet/src/errors.ts [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+
+export class ArrayNoSameLength extends Error {
+    constructor(first: string, second: string) {
+        super(`${first} and ${second} must have the same length`);
+    }
+}
+
+export class NegativeTokenIdError extends Error {
+    constructor(tokenId: bigint, amount: bigint) {
+        super(`negative value for tokenId ${tokenId}: ${amount}`);
```
Comment (tuler) [inline packages/wallet/src/errors.ts:47]: are we raising this in what situation? tokenId is never negative as it comes from a uint256

File: packages/wallet/src/errors.ts [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+
+export class TokenFromUserNotFound extends Error {
+    constructor(user: string, token: string, tokenId: bigint) {
+        super(`user ${user} does not have token ${tokenId} of token ${token}`);
+    }
+}
+
+export class ArrayNoSameLength extends Error {
+    constructor(first: string, second: string) {
+        super(`${first} and ${second} must have the same length`);
```
Comment (tuler) [inline packages/wallet/src/errors.ts:41]: are we raising this in what situation?

File: packages/wallet/src/errors.ts [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+
+export class NegativeTokenIdError extends Error {
+    constructor(tokenId: bigint, amount: bigint) {
+        super(`negative value for tokenId ${tokenId}: ${amount}`);
+    }
+}
+
+export class ArrayEmptyError extends Error {
+    constructor(name: string) {
+        super(`${name} must not be empty`);
```
Comment (tuler) [inline packages/wallet/src/errors.ts:53]: are we raising this in what situation?

File: packages/wallet/src/wallet.ts [hunk 16]
```diff
@@ -1,28 +1,35 @@
...
+
+import { ether, erc20, erc721, erc1155Single, erc1155Batch } from "./contracts";
+
+import { depositHandler } from "./token";
+import { RelayError } from "./errors";
 
 export type Wallet = {
     ether: bigint;
-    erc20: Record<Address, bigint>;
+    erc20: Record<string, bigint>;
```
Comment (tuler) [inline packages/wallet/src/wallet.ts:11]: is the string in this case required to be an Address?
If yes, it should be an Address type.

Comment (sandhilt) [reply packages/wallet/src/wallet.ts:11]: Some address can be alias, but in some cases, like withdraw to generate voucher, need to be a Address type.
So, when you need this, transfer for a real address, and call withdraw method.
This alias follow this pattern: https://github.com/tuler/deroll/blob/903d8ec54ef7ba39a127b8cb4e61879a495be4f8/packages/wallet/src/wallet.ts#L58

**Comment by sandhilt** [comment]
> I still don't understand why we need the design of `DepositOperation`, `DepositArgs`, some of the error classes. I think 
> ```
> 
> And each token module exports a handler function on its own.

Ok, I can follow this pattern.
And how I access field `wallet` from `WalletImpl` inside `handler` (method or function of module) based in `AdvanceRequestHandler`?

File: packages/wallet/src/errors.ts (hunk 11 above)
Comment (sandhilt) [reply packages/wallet/src/errors.ts:29]: In this situation, the user is asking to generate a voucher from a non-existent wallet. I can also change to generate an empty wallet. What do you think @fabiooshiro ?

File: packages/wallet/src/index.ts [hunk 17]
```diff
@@ -61,8 +88,83 @@ export const parseERC20Deposit = (payload: Payload): ERC20Deposit => {
...
 
+/**
+ * Decode input according to https://github.com/cartesi/rollups/tree/v1.2.0#input-encodings-for-deposits
+ * @param payload input payload
+ * @returns
+ */
+export const parseERC721Deposit = (payload: Payload): ERC721Deposit => {
+    const token = getAddress(slice(payload, 0, 20)); // 20 bytes for address
+    const sender = getAddress(slice(payload, 20, 40)); // 20 bytes for address
+    const commonPayload = slice(payload, 40);
```
Comment (tuler) [inline packages/wallet/src/index.ts:99]: This parsing is not what documentation says

File: packages/wallet/src/index.ts [hunk 18]
```diff
@@ -61,8 +88,83 @@ export const parseERC20Deposit = (payload: Payload): ERC20Deposit => {
     return { success, token, sender, amount };
 };
 
+/**
+ * Decode input according to https://github.com/cartesi/rollups/tree/v1.2.0#input-encodings-for-deposits
```
Comment (tuler) [inline packages/wallet/src/index.ts:92]: fix link

File: packages/wallet/src/errors.ts (hunk 15 above)
Comment (sandhilt) [reply packages/wallet/src/errors.ts:53]: see: https://github.com/tuler/deroll/pull/62#discussion_r1546686258

File: packages/wallet/src/errors.ts (hunk 14 above)
Comment (sandhilt) [reply packages/wallet/src/errors.ts:41]: see: https://github.com/tuler/deroll/pull/62#discussion_r1546686258

File: packages/wallet/src/errors.ts (hunk 13 above)
Comment (sandhilt) [reply packages/wallet/src/errors.ts:47]: Renaming for when the values passed are negative in the `amount` and not in the `tokenId`.

File: packages/wallet/src/errors.ts (hunk 12 above)
Comment (sandhilt) [reply packages/wallet/src/errors.ts:35]: Ok removed.

File: packages/wallet/src/index.ts (hunk 17 above)
Comment (sandhilt) [reply packages/wallet/src/index.ts:99]: Can you explain more about this? Does this refers to the `baseLayerData` and `execLayerData`?

File: packages/wallet/src/errors.ts (hunk 11 above)
Comment (sandhilt) [reply packages/wallet/src/errors.ts]: Ok removed.

**Pull request closed without merging**